service: query_pager: fix last-position for filtering queries

denesb opened the pull request:
On short-pages, cut short because of a tombstone prefix. When page-results are filtered and the filter drops some rows, the last-position is taken from the page visitor, which does the filtering. This means that last partition and row position will be that of the last row the filter saw. This will not match the last position of the replica, when the replica cut the page due to tombstones. When fetching the next page, this means that all the tombstone suffix of the last page, will be re-fetched. Worse still: the last position of the next page will not match that of the saved reader left on the replica, so the saved reader will be dropped and a new one created from scratch.
This wasted work will show up as elevated tail latencies. Fix by always taking the last position from raw query results.

Fixes: #22620

This bug exists since the introduction of empty replica pages, so all live versions need a backport.

avikivity commented:
Wow

avikivity commented:
Goes back to Scylla 0.12

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14881/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3106/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14238/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6226/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14881/)
#### Failed Tests (43/36877):
- [alternator.test_scan.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.test_scan.debug.1)
- [alternator.test_scan.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.test_scan.debug.2)
- [alternator.test_scan.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.test_scan.debug.3)
- [alternator.test_scan.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.test_scan.release.1)
- [alternator.test_scan.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.test_scan.release.2)
- [alternator.test_scan.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.test_scan.release.3)
- [test_scan_long_partition_tombstone_string.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/alternator/test_scan/Tests___Unit_Tests___test_scan_long_partition_tombstone_string_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scan_long_partition_tombstone_string.debug.1)
- [test_scan_long_partition_tombstone_string.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/alternator/test_scan/Tests___Unit_Tests___test_scan_long_partition_tombstone_string_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scan_long_partition_tombstone_string.debug.2)
- [test_scan_long_partition_tombstone_string.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/alternator/test_scan/Tests___Unit_Tests___test_scan_long_partition_tombstone_string_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scan_long_partition_tombstone_string.debug.3)
- [test_scan_long_partition_tombstone_string.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14881/testReport/junit/alternator/test_scan/Tests___Unit_Tests___test_scan_long_partition_tombstone_string_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scan_long_partition_tombstone_string.dev.2)
#### Build Details:
- Duration: 20 hr
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14893/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6234/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3114/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14246/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14893/)
#### Failed Tests (12/37824):
- [test_cluster_reduction_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/cdc_test/TestCdc/Tests___dtest___test_cluster_reduction_with_cdc_Multi_DC_cluster_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_reduction_with_cdc[Multi_DC_cluster])
- [test_cluster_reduction_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/cdc_test/TestCdc/Tests___dtest_with_topology_changes___test_cluster_reduction_with_cdc_Multi_DC_cluster_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_reduction_with_cdc[Multi_DC_cluster])
- [test_switch_with_password_auth](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/ldap_test/TestLdapSaslAuth/Tests___dtest_with_tablets___test_switch_with_password_auth) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_switch_with_password_auth)
- [test_switch_with_password_auth](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/ldap_test/TestLdapSaslAuth/Tests___dtest_with_tablets___test_switch_with_password_auth_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_switch_with_password_auth)
- [test_switch_with_password_auth](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/ldap_test/TestLdapSaslAuth/Tests___dtest___test_switch_with_password_auth) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_switch_with_password_auth)
- [test_index_drop_add](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/lwt_schema_modification_test/TestLWTSchemaModification/Tests___dtest_with_topology_changes___test_index_drop_add) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_drop_add)
- [test_writes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/per_partition_rate_limiting_test/TestPerPartitionRateLimiting/Tests___dtest_with_topology_changes___test_writes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_writes)
- [test_ignore_dead_nodes_for_replace_option_multi_rack](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest___test_ignore_dead_nodes_for_replace_option_multi_rack) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ignore_dead_nodes_for_replace_option_multi_rack)
- [test_ignore_dead_nodes_for_replace_option_multi_rack](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest___test_ignore_dead_nodes_for_replace_option_multi_rack_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ignore_dead_nodes_for_replace_option_multi_rack)
- [test_safe_unsafe_rebuild[2]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14893/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest___test_safe_unsafe_rebuild_2_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_safe_unsafe_rebuild[2])
#### Build Details:
- Duration: 4 hr 46 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14894/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3116/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6236/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14248/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14894/)
#### Failed Tests (18/37844):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/boost.encryption_at_rest_test.test_kms_network_error.2.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [test_cluster_reduction_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/cdc_test/TestCdc/Tests___dtest_with_topology_changes___test_cluster_reduction_with_cdc_Multi_DC_cluster_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_reduction_with_cdc[Multi_DC_cluster])
- [test_increment_decrement_counters_in_threads](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/counter_tests/TestCounters/Tests___dtest_with_topology_changes___test_increment_decrement_counters_in_threads) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_increment_decrement_counters_in_threads)
- [test_cast_smallint](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/cql_cast_test/TestCQLCast/Tests___dtest_with_tablets___test_cast_smallint) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cast_smallint)
- [test_data_distribution_balance[6-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_topology_changes___test_data_distribution_balance_6_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[6-TimeWindowCompactionStrategy])
- [test_data_distribution_balance[6-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/data_distribution_balance_test/TestDataDistribution/Tests___dtest_with_topology_changes___test_data_distribution_balance_6_TimeWindowCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_data_distribution_balance[6-TimeWindowCompactionStrategy])
- [test_reboot[local]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest___test_reboot_local_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_reboot[local])
- [test_reboot[replicated]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest___test_reboot_replicated_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_reboot[replicated])
- [test_reboot[replicated]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest___test_reboot_replicated__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_reboot[replicated])
- [test_hintedhandoff_decom](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14894/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_decom) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_decom)
#### Build Details:
- Duration: 8 hr 5 min
- Builder: spider4.cloudius-systems.com

avikivity commented:
The test_data_distribution failure is similar to what I analyzed in https://github.com/scylladb/scylladb/pull/22594

denesb commented:
New in v2:
* Added reproducer test which revealed that the bug is less dangerous than I thought: it only affects short pages cut short due to tombstone suffix.
* Adjusted description of the patch according to the above.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14971/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14314/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3178/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6298/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14971/)
#### Failed Tests (1/37875):
- [cqlpy.cassandra_tests/validation/entities/frozen_collections_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14971/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_frozen_collections_test_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/frozen_collections_test.debug.2)
#### Build Details:
- Duration: 10 hr
- Builder: i-0c23ecc38a0702b24 (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: UNSTABLE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14986/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3194/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14329/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6314/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14986/)
#### Build Details:
- Duration: 8 hr 15 min
- Builder: spider2.cloudius-systems.com

denesb commented:
> [cqlpy.cassandra_tests/validation/entities/frozen_collections_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14971/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_frozen_collections_test_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/frozen_collections_test.debug.2)

https://github.com/scylladb/scylladb/issues/21292
